library/cwm/src/lib/cwm/widget.rb

```diff
@@ -0,0 +1,425 @@
+    #       [ "USA", _("United States of America")],
+    #       [ "North Pole", _("Really cold place")],
+    #     ]
+    #   end
+    #
+    def description
+      unless respond_to?(:label)
+        raise "For input field widget '#{self.class}' label method have to be defined"
+      end
+
+      res = {
+        "widget"        => :multi_selection_box,
+      }
+      res["items"] = items if respond_to?(:items)
+
+      res.merge(super)
+    end
+
+  protected
+    def values
+      Yast::UI.QueryWidget(Id(widget_id), :SelectedItems)
+    end
+
+    def values=(val)
+      Yast::UI.ChangeWidget(Id(widget_id), :SelectedItems, val)
+    end
+  end
+
+  class IntField < AbstractionWidget
+    include ValueBasedWidget
+
+    # description for combobox additionally support `minimum` and `maximum` methods.
+    # Both methods have to FixNum, where it is limited by C signed int range (-2**30 to 2**31-1).
+    # @example minimum and maximum methods
+    #   def minimum
+    #     50
+    #   end
+    #
+    #   def maximum
+    #     200
+    #   end
+    #
+    def description
+      unless respond_to?(:label)
+        raise "For input field widget '#{self.class}' label method have to be defined"
+      end
+
+      res = {
+        "widget"        => :intfield,
+      }
+      res["minimum"] = minimum if respond_to?(:minimum)
+      res["maximum"] = minimum if respond_to?(:maximum)
+
+      res.merge(super)
+    end
+  end
+
+  class RadioButtonsWidget < AbstractWidget
+    # description for radio buttons additionally support `items` method.
+    # `items` method have to return array of string pairs, where first value is
+    # radio button id and second is button label.
+    # @example items method
+    #   def items
+    #     [
+    #       [ "Canada", _("Canada")],
+    #       [ "USA", _("United States of America")],
+    #       [ "North Pole", _("Really cold place")],
+    #     ]
+    #   end
+    #
+    def description
+      unless respond_to?(:label)
+        raise "For radio buttons widget '#{self.class}' label method have to be defined"
+      end
+
+      res = {
+        "widget"        => :radio_buttons, # yeah, really so incosistent to have combox and selection_box
+      }
+      res["items"] = items if respond_to?(:items)
+
+      res.merge(super)
+    end
+
+  protected
+    def value
+      Yast::UI.QueryWidget(Id(widget_id), :CurrentButton)
+    end
+
+    def value=(val)
+      Yast::UI.ChangeWidget(Id(widget_id), :CurrentButton, val)
+    end
+  end
+
+  class PushButtonWidget < AbstractWidget
+    # @param [Symbol] id is used as widget id
+    # @yield block runned after button is clicked
+    def initialize(id, &block)
+      self.widget_id = id
+      @block = block
+    end
+
+    def description
+      {
+        "widget"        => :push_button
+      }.merge(super)
+    end
+
+  private
+
+    def handle(widget, event)
+      return if event["ID"] != widget
+
+      @block.call
+
+      nil
```

Comment:
`CWM handle` function docs say

> It should return nil to continue processing or a symbol to be returned from the dialog.

If you differ intentionally, how do you intend to exit the event loop?